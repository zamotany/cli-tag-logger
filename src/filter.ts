import stripAnsi from 'strip-ansi';

export type LevelTag = 'debug' | 'info' | 'success' | 'warn' | 'error' | RegExp;

export type FilterOptions = {
  only?: LevelTag | LevelTag[];
  exclude?: LevelTag | LevelTag[];
};

export type Tester = (message: string) => boolean;

export type FilterConfig = FilterOptions | Tester;

function toArray<T>(value: T | T[]): T[] {
  return Array.isArray(value) ? value : [value];
}

export function createTester(config: FilterConfig = {}): Tester {
  if (typeof config === 'function') {
    return message => config(stripAnsi(message));
  }

  if (config.exclude === undefined && config.only === undefined) {
    return () => true;
  }

  return message => {
    const value = stripAnsi(message);

    let include = toArray(config.only || [/.*/]).some(spec => {
      if (typeof spec === 'string') {
        return value.startsWith(spec);
      }

      return spec.test(value);
    });

    if (!include) {
      return false;
    }

    include = !toArray(config.exclude || []).some(spec => {
      if (typeof spec === 'string') {
        return value.startsWith(spec);
      }

      return spec.test(value);
    });

    return include;
  };
}
