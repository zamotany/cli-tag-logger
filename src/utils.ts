import utils from 'util';

export function stringify(value: any): string {
  return typeof value === 'string'
    ? value
    : utils.inspect(value, {
        depth: null,
        maxArrayLength: Infinity,
        colors: !process.env.NO_COLOR,
      });
}

export function mergeStringsWithValues(
  prefix: (() => string) | undefined,
  strings: TemplateStringsArray,
  values: any[]
): string {
  const stringifiedValues = values.map(stringify);

  const segments: string[] = new Array(
    strings.length + stringifiedValues.length + (prefix ? 1 : 0)
  );

  if (prefix) {
    segments[0] = prefix();
  }

  for (let i = 0, j = prefix ? 1 : 0; i < strings.length; i++) {
    segments[j++] = strings[i];
    if (stringifiedValues[i]) {
      segments[j++] = stringifiedValues[i];
    }
  }

  return segments.join('');
}

export type ComposableValues = Array<
  string | number | false | null | undefined | object
>;

export function compose(...values: ComposableValues): string {
  return values
    .map(value => {
      if (typeof value === 'string' || typeof value === 'number') {
        return value.toString();
      }

      if (value === false || value === null || value === undefined) {
        return '';
      }

      return stringify(value);
    })
    .filter(Boolean)
    .join(' ');
}
