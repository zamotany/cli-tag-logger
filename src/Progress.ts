import { ComposableValues, compose } from './utils';
import restoreCursor from 'restore-cursor';
import { terminal } from 'terminal-kit';

type PrintFn = (...values: ComposableValues) => void;

type ProgressConfig = {
  renderBottom: (print: PrintFn, info: { delta: number }) => void;
};

class Progress {
  lastTickTimestamp: number = Date.now();
  topY: number = 0;
  bottomHeight: number = 0;
  renderBottom: ProgressConfig['renderBottom'];

  constructor(config: ProgressConfig) {
    this.renderBottom = config.renderBottom;
    terminal.getCursorLocation((error, x, y) => {
      if (error || x === undefined || y === undefined) {
        throw new Error('Cannot get cursor position');
      }
      this.topY = y;
      terminal.hideCursor(true);
      this.tick();
    });
  }

  tick = () => {
    debugger;
    const now = Date.now();
    const delta = Math.abs(now - this.lastTickTimestamp);
    this.lastTickTimestamp = now;
    const lines: string[] = [];
    this.renderBottom(
      (...values) => {
        lines.push(...compose(...values).split('\n'));
      },
      { delta }
    );

    this.bottomHeight = lines.length;

    // Lines of available area
    const fillLength = Math.max(0, terminal.height - this.topY);

    terminal.moveTo(0, this.topY);
    terminal.eraseDisplayBelow();
    terminal('\n'.repeat(fillLength));
    // If there's enough room for bottom part, the cursor needs to be moved up.
    // Otherwise, we need to update topY since rendering bottom section will push new lines.
    if (fillLength + 1 >= this.bottomHeight) {
      terminal.moveTo(0, terminal.height - this.bottomHeight + 1);
    } else {
      this.topY -= this.bottomHeight - 1;
    }
    terminal(lines.join('\n'));
  };

  printAbove: PrintFn = (...values) => {
    debugger;
    const lines = compose(...values).split('\n');
    terminal.moveTo(0, this.topY);
    terminal.eraseDisplayBelow();
    terminal(lines.join('\n'));
    this.topY += lines.length;
    this.tick();
  };
}

export function makeProgress(config: ProgressConfig): Progress {
  restoreCursor();
  return new Progress(config);
}

const { tick, printAbove } = makeProgress({
  renderBottom: (print, { delta }) => {
    print('hello');
    print('world', delta);
    // if (Math.random() > 0.5) {
    //   print('random');
    // }
  },
});

setInterval(() => {
  // tick();
}, 1000);

setInterval(() => {
  printAbove('print above', Date.now());
}, 1000);
