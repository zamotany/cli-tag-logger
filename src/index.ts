import utils from 'util';
import colorette from 'colorette';
import { EventEmitter } from 'events';
import path from 'path';
import fs from 'fs';
import stripAnsi from 'strip-ansi';

const coloretteStyles = Object.entries(colorette).filter(
  ([, value]) => typeof value === 'function'
) as Array<[string, colorette.Style]>;

function stringify(value: any): string {
  return typeof value === 'string'
    ? value
    : utils.inspect(value, {
        depth: null,
        maxArrayLength: Infinity,
        colors: !process.env.NO_COLOR,
      });
}

function mergeStringsWithValues(
  prefix: (() => string) | undefined,
  strings: TemplateStringsArray,
  values: any[]
): string {
  const stringifiedValues = values.map(stringify);

  const segments: string[] = new Array(
    strings.length + stringifiedValues.length + (prefix ? 1 : 0)
  );

  if (prefix) {
    segments[0] = prefix();
  }

  for (let i = 0, j = prefix ? 1 : 0; i < strings.length; i++) {
    segments[j++] = strings[i];
    if (stringifiedValues[i]) {
      segments[j++] = stringifiedValues[i];
    }
  }

  return segments.join('');
}

export const styles = coloretteStyles.reduce((acc, [key, style]) => {
  return {
    ...acc,
    [key]: (strings: TemplateStringsArray, ...values: any[]) => {
      colorette.options.enabled = !process.env.NO_COLOR;
      return style(mergeStringsWithValues(undefined, strings, values));
    },
  };
}, {} as { [K in keyof Omit<typeof colorette, 'options'>]: typeof String.raw });

export function createTag(prefix?: () => string) {
  return (strings: TemplateStringsArray, ...values: any[]): string => {
    return mergeStringsWithValues(prefix, strings, values);
  };
}

export const debug = createTag(() => styles.gray`debug `);
export const info = createTag(() => styles.blue`info `);
export const warn = createTag(() => styles.yellow`warn `);
export const error = createTag(() => styles.red`error `);
export const success = createTag(() => styles.green`success `);
export const inspect = createTag();
export const trace = (
  strings: TemplateStringsArray,
  ...values: any[]
): string => {
  const err = {
    name: '',
    stack: undefined,
  };
  Error.captureStackTrace(err, trace);
  return compose(
    mergeStringsWithValues(() => styles.magenta`trace `, strings, values),
    err.stack
  );
};

export type ComposableValues = Array<
  string | number | false | null | undefined | object
>;

export function compose(...values: ComposableValues): string {
  return values
    .map(value => {
      if (typeof value === 'string' || typeof value === 'number') {
        return value.toString();
      }

      if (value === false || value === null || value === undefined) {
        return '';
      }

      return stringify(value);
    })
    .filter(Boolean)
    .join(' ');
}

const emitter = new EventEmitter();
const consolePrint = (message: any) => {
  console.log(message);
};
emitter.on('print', consolePrint);

export type Configuration = {
  // verbose?: boolean;
  console?: boolean;
  file?: string;
  json?: boolean;
  listener?: (type: 'print', message: string) => void;
};

export function configure(options: Configuration): () => void {
  const {
    console: logToConsole = true,
    file,
    json,
    // verbose,
    listener,
  } = options;
  let dispose = () => {};

  if (!logToConsole) {
    emitter.removeListener('print', consolePrint);
  }

  if (file) {
    const filename = path.isAbsolute(file) ? file : path.resolve(file);
    const fileListener = (message: string) => {
      const data = json
        ? JSON.stringify({ timestamp: new Date(), message: stripAnsi(message) })
        : `${new Date().toJSON()} ${stripAnsi(message)}`;
      fs.appendFileSync(filename, data + '\n');
    };
    emitter.on('print', fileListener);
    dispose = () => {
      emitter.off('print', fileListener);
    };
  }

  if (listener) {
    emitter.on('print', listener);
  }

  return dispose;
}

export function print(...values: ComposableValues) {
  emitter.emit('print', compose(...values));
}
