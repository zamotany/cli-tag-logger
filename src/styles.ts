import colorette from 'colorette';
import { mergeStringsWithValues } from './utils';

const coloretteStyles = Object.entries(colorette).filter(
  ([, value]) => typeof value === 'function'
) as Array<[string, colorette.Style]>;

export const styles = coloretteStyles.reduce((acc, [key, style]) => {
  return {
    ...acc,
    [key]: (strings: TemplateStringsArray, ...values: any[]) => {
      colorette.options.enabled = !process.env.NO_COLOR;
      return style(mergeStringsWithValues(undefined, strings, values));
    },
  };
}, {} as { [K in keyof Omit<typeof colorette, 'options'>]: typeof String.raw });
