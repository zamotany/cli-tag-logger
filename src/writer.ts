import path from 'path';
import fs from 'fs-extra';
import stripAnsi from 'strip-ansi';
import onExit from 'signal-exit';

import { compose, ComposableValues } from './utils';
import { FilterConfig, Tester, createTester } from './filter';

export abstract class Writer {
  private tester: Tester;

  constructor({ filter }: { filter?: FilterConfig } = {}) {
    this.tester = createTester(filter);
  }

  onPrint(_message: string) {
    throw new Error('onPrint method from Writer class must be implemented');
  }

  print = (...values: ComposableValues) => {
    const message = compose(...values);
    if (this.tester(message)) {
      this.onPrint(message);
    }
  };
}

export function composeWriters<
  MainWriter extends Writer,
  OtherWriter extends Writer
>(main: MainWriter, ...writers: OtherWriter[]) {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const { print, onPrint: _onPrint, ...exposedValues } = main;

  // Filter out functions exposed on MainWriter
  type ExposedValues = typeof exposedValues;
  const exposedFunctions: {
    [FnKey in keyof ExposedValues]: ExposedValues[FnKey] extends Function
      ? ExposedValues[FnKey]
      : never;
  } = Object.entries<unknown>(exposedValues)
    .filter(([, value]) => typeof value === 'function')
    .reduce((acc, [key, fn]) => ({ ...acc, [key]: fn }), {}) as any;

  return {
    print: (...values: ComposableValues) => {
      const message = compose(...values);
      print(message);
      writers.forEach(writer => writer.print(message));
    },
    ...exposedFunctions,
  };
}

export class ConsoleWriter extends Writer {
  constructor({ filter }: { filter?: FilterConfig } = {}) {
    super({ filter });
  }

  onPrint(message: string) {
    console.log(message);
  }
}

export class FileWriter extends Writer {
  private fd?: number;
  public readonly json: boolean;

  constructor(
    public readonly filename: string,
    { filter, json }: { filter?: FilterConfig; json?: boolean } = {}
  ) {
    super({ filter });
    this.json = Boolean(json);
    this.filename = path.isAbsolute(filename)
      ? filename
      : path.resolve(filename);
    fs.ensureFileSync(this.filename);
    this.fd = fs.openSync(this.filename, 'a');
    onExit(this.close);
  }

  close = () => {
    this.fd !== undefined && fs.closeSync(this.fd);
    this.fd = undefined;
  };

  onPrint(message: string) {
    const data = this.json
      ? JSON.stringify({
          timestamp: new Date(),
          message: stripAnsi(message),
        })
      : `${new Date().toJSON()} ${stripAnsi(message)}`;
    this.fd !== undefined && fs.appendFileSync(this.fd, data + '\n');
  }
}
