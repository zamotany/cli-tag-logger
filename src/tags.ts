import { mergeStringsWithValues, compose } from './utils';
import { styles } from './styles';

export function createTag(prefix?: () => string) {
  return (strings: TemplateStringsArray, ...values: any[]): string => {
    return mergeStringsWithValues(prefix, strings, values);
  };
}

export const debug = createTag(() => styles.gray`debug `);
export const info = createTag(() => styles.blue`info `);
export const warn = createTag(() => styles.yellow`warn `);
export const error = createTag(() => styles.red`error `);
export const success = createTag(() => styles.green`success `);
export const inspect = createTag();
export const trace = (
  strings: TemplateStringsArray,
  ...values: any[]
): string => {
  const err = {
    name: '',
    stack: undefined,
  };
  Error.captureStackTrace(err, trace);
  return compose(
    mergeStringsWithValues(() => styles.magenta`trace `, strings, values),
    err.stack
  );
};
