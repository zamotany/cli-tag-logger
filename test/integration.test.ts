import * as log from '../src';

describe('cli-tag-logger', () => {
  beforeAll(() => {
    process.env.NO_COLOR = '1';
  });

  it('should provide styling tags', () => {
    expect(log.styles).toMatchSnapshot();
    expect(log.styles.red`hello`).toEqual('hello');
  });

  it('should provide debug, info, success, warn, error tags', () => {
    expect(log.debug`hello ${1}`).toEqual('debug hello 1');
    expect(log.info`hello ${{ a: 1 }}`).toEqual('info hello { a: 1 }');
    expect(log.success`hello ${/regex/}`).toEqual('success hello /regex/');
    expect(log.warn`hello ${function a() {}}`).toEqual(
      'warn hello [Function: a]'
    );
    expect(log.error`hello`).toEqual('error hello');
  });

  it('should provie trace and inspect tags', () => {
    expect(log.inspect`hello ${{ a: 1 }}`).toEqual('hello { a: 1 }');
    expect(log.trace`hello`).toMatch(/trace hello Error:(\s*at .+\s*)+/);
  });

  it('should allow to compose messages', () => {});

  it('should print message', () => {});

  it('should allow printing to file', () => {});

  it('should allow printing to file as json', () => {});
});
