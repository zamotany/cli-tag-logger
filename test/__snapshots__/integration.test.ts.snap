// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cli-tag-logger should provide styling tags 1`] = `
Object {
  "bgBlack": [Function],
  "bgBlackBright": [Function],
  "bgBlue": [Function],
  "bgBlueBright": [Function],
  "bgCyan": [Function],
  "bgCyanBright": [Function],
  "bgGreen": [Function],
  "bgGreenBright": [Function],
  "bgMagenta": [Function],
  "bgMagentaBright": [Function],
  "bgRed": [Function],
  "bgRedBright": [Function],
  "bgWhite": [Function],
  "bgWhiteBright": [Function],
  "bgYellow": [Function],
  "bgYellowBright": [Function],
  "black": [Function],
  "blackBright": [Function],
  "blue": [Function],
  "blueBright": [Function],
  "bold": [Function],
  "cyan": [Function],
  "cyanBright": [Function],
  "dim": [Function],
  "gray": [Function],
  "green": [Function],
  "greenBright": [Function],
  "hidden": [Function],
  "inverse": [Function],
  "italic": [Function],
  "magenta": [Function],
  "magentaBright": [Function],
  "red": [Function],
  "redBright": [Function],
  "reset": [Function],
  "strikethrough": [Function],
  "underline": [Function],
  "white": [Function],
  "whiteBright": [Function],
  "yellow": [Function],
  "yellowBright": [Function],
}
`;
